pennylane/devices/qutrit_mixed/sampling.py

```diff
@@ -0,0 +1,334 @@
+    if not mp_has_obs:
+        observables = ["".join(observable.astype("str")) for observable in observables]
+    return dict(zip(observables, counts))
+
+
+def _measure_with_samples_diagonalizing_gates(
+    mps: List[Union[SampleMP, CountsMP]],
+    state: np.ndarray,
+    shots: Shots,
+    is_state_batched: bool = False,
+    rng=None,
+    prng_key=None,
+) -> TensorLike:
+    """Returns the samples of the measurement process performed on the given state,
+    by rotating the state into the measurement basis using the diagonalizing gates
+    given by the measurement process.
+
+    Args:
+        mp (~.measurements.SampleMeasurement): The sample measurement to perform
+        state (np.ndarray[complex]): The state vector to sample from
+        shots (~.measurements.Shots): The number of samples to take
+        is_state_batched (bool): whether the state is batched or not
+        rng (Union[None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``.
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. Only for simulation using JAX.
+
+    Returns:
+        TensorLike[Any]: Sample measurement results
+    """
+    # apply diagonalizing gates
+    state = _apply_diagonalizing_gates(mps, state, is_state_batched)
+
+    total_indices = get_num_wires(state, is_state_batched)
+    wires = qml.wires.Wires(range(total_indices))
+
+    def _process_single_shot(samples):
+        processed = []
+        for mp in mps:
+            if isinstance(mp, SampleMP):
+                res = mp.process_samples(samples, wires)
+                res = math.squeeze(res)
+            elif isinstance(mp, CountsMP):
+                res = _process_counts_samples(mp, samples, wires)
+            else:
+                raise NotImplementedError
+            processed.append(res)
+        return tuple(processed)
+
+    # if there is a shot vector, build a list containing results for each shot entry
+    if shots.has_partitioned_shots:
+        processed_samples = []
+        for s in shots:
+            # Like default.qubit currently calling sample_state for each shot entry,
+            # but it may be better to call sample_state just once with total_shots,
+            # then use the shot_range keyword argument
+            samples = sample_state(
+                state,
+                shots=s,
+                is_state_batched=is_state_batched,
+                wires=wires,
+                rng=rng,
+                prng_key=prng_key,
+            )
+            processed_samples.append(_process_single_shot(samples))
+
+        return tuple(zip(*processed_samples))
+
+    samples = sample_state(
```

Comment:
I noticed in the qubit device that this call is in a try/except block to handle cases where you end up with NaNs in the probabilities. Is that avoided here somehow ?

Reply:
the NaNs are for mid-circuit measurements, I can re-add that, but I thought simplicity would be better for now as we removed the NaN tests in measure.

The tests in `measure` could be re-added to measure, they were removed to use `math.clip` instead of `math.where` for removing negative probability values since `math.clip` didn't work with NaNs. We reverted to `math.where` for a separate interface issue.